design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,212 @@
+# Deprecating old Ceph OSD types (May 2019)
+**Targeted for v1.1-v1.2**
+
+## Background
+Rook's current implementation for preparing and starting OSDs includes several different patterns
+that have evolved over Rook's lifetime. These patterns include different code paths for supporting
+directories and devices as storage targets, and further separated into support for both Filestore
+and Bluestore as backing stores. There is further complication added by support for `ceph-volume` as
+a tool for storage provisioning versus continued support for legacy installations of Rook which used
+manual partitioning.
+
+There has been talk for quite some time about removing support for directories as storage targets,
+as this is a use case in Ceph that is not supported for production. Additionally, Filestore is
+effectively in sustaining mode, and Bluestore is the defacto replacement for it.
+
+All of the various OSD support patterns/code-paths need changes to allow the [planned updates to
+Rook's configuration strategy](/design/ceph-config-updates.md) to proceed. The various code paths
+are intertangled and difficult to separate, and this presents a huge hurdle for the config changes.
+In general, the intertangled OSD code poses a risk for long-term maintenance, as it may be more
+prone to bugs being introduced, and it has and will continue to slow development.
+
+In light of this background information, this design proposes three high-level goals:
+1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some form
+2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
+and that is extensible for the future.
+3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
+   deployments which use old methods to begin using the new methods.
+
+
+## Deprecated features and code paths
+This design proposes to deprecate ...
+- the use of directories as storage targets
```

Comment:
@travisn and I had a conversation a few weeks back about still including support for a directory for bluestore, where we just create a file for the ``block`` file/device.  This is reasonably simple and (from a ceph perspective) will be suppoortable going forward.  It does mean one non-ceph-volume OSD path, though.

Reply:
Yeah. Travis mentioned that this morning, and I haven't considered how to include it in this design. We did talk about effectively making that a "hidden" option that isn't exposed to users.

Reply:
What if we had a simple CRD property that indicates how big of an OSD to create under dataDirHostPath? I don't think we can make it hidden, but it needs to be obviously a non-production setting.
```
  dataDirHostPath: /var/lib/rook
  dataDirHostPathStorageForTestEnv: 5Gi
```

Or perhaps that property is just a bool, and we could automatically choose a size from 1-10Gi depending on available free space, but the size matters since we have to reserve it up front.

Reply:
Wouldn't it still be possible to add an env var that the operator could read to the same effect? `TESTING_STORAGE_FILE_SIZE_GiB`. We could assume not present or less than one means do-not-create.

This would work fine for simple dev and test, but we would be failing to exercise a lot of OSD code if we don't test in environments with actual disks. We could also look into creating virtual disks, but I vaguely recall wanting to do that in the past and finding that it was not a straightforward process.